src/main/java/frc/robot/IMU/BNO055.java

```diff
@@ -0,0 +1,776 @@
+			return val;
+		}
+	}
+
+	public class RevInfo {
+		public byte accel_rev;
+		public byte mag_rev;
+		public byte gyro_rev;
+		public short sw_rev;
+		public byte bl_rev;
+	}
+
+	public class CalData {
+		public byte sys;
+		public byte gyro;
+		public byte accel;
+		public byte mag;
+	}
+
+	public enum vector_type_t {
+		VECTOR_ACCELEROMETER (reg_t.BNO055_ACCEL_DATA_X_LSB_ADDR.getVal()),
+		VECTOR_MAGNETOMETER  (reg_t.BNO055_MAG_DATA_X_LSB_ADDR.getVal()),
+		VECTOR_GYROSCOPE     (reg_t.BNO055_GYRO_DATA_X_LSB_ADDR.getVal()),
+		VECTOR_EULER         (reg_t.BNO055_EULER_H_LSB_ADDR.getVal()),
+		VECTOR_LINEARACCEL   (reg_t.BNO055_LINEAR_ACCEL_DATA_X_LSB_ADDR.getVal()),
+		VECTOR_GRAVITY       (reg_t.BNO055_GRAVITY_DATA_X_LSB_ADDR.getVal());
+
+		private final int val;
+
+		vector_type_t(int val) {
+			this.val = val;
+		}
+
+		public int getVal() {
+			return val;
+		}
+	};
+	
+	/**
+	 * Instantiates a new BNO055 class.
+	 *
+	 * @param port the physical port the sensor is plugged into on the roboRio
+	 * @param address the address the sensor is at (0x28 or 0x29)
+	 */
+	private BNO055(I2C.Port port, byte address) {
+		imu = new I2C(port, address);
+		
+		executor = new java.util.Timer();
+		executor.schedule(new BNO055UpdateTask(this), 0L, THREAD_PERIOD);
+	}
+
+	/**
+	 * Get an instance of the IMU object.
+	 * 
+	 * @param mode the operating mode to run the sensor in.
+	 * @param port the physical port the sensor is plugged into on the roboRio
+	 * @param address the address the sensor is at (0x28 or 0x29)
+	 * @return the instantiated BNO055 object
+	 */
+	public static BNO055 getInstance(opmode_t mode, vector_type_t vectorType,
+			I2C.Port port, byte address) {
+		if(instance == null) {
+			instance = new BNO055(port, address);
+		}
+		requestedMode = mode;
+		requestedVectorType = vectorType;
+		return instance;
+	}
+
+	/**
+	 * Get an instance of the IMU object plugged into the onboard I2C header.
+	 *   Using the default address (0x28)
+	 *  
+	 * @param mode the operating mode to run the sensor in.
+	 * @param vectorType the format the position vector data should be returned
+	 *   in (if you don't know use VECTOR_EULER).
+	 * @return the instantiated BNO055 object
+	 */
+	public static BNO055 getInstance(opmode_t mode, vector_type_t vectorType) {
+		return getInstance(mode, vectorType, I2C.Port.kOnboard,
+				BNO055_ADDRESS_A);
+	}
+
+
+	/**
+	 * Called periodically. Communicates with the sensor, and checks its state. 
+	 */
+	private void update() {
+		currentTime = Timer.getFPGATimestamp(); //seconds
+		if(!initialized) {
+//			System.out.println("State: " + state + ".  curr: " + currentTime
+//					+ ", next: " + nextTime);
+			
+			//Step through process of initializing the sensor in a non-
+			//  blocking manner. This sequence of events follows the process
+			//  defined in the original adafruit source as closely as possible.
+			//  XXX: It's likely some of these delays can be optimized out.
+			switch(state) {
```

Comment:
BTW, this is one way to implement a State Machine.  Not a very readable way, they could use enum's instead of int's.  Or at least constants to hint what's going on (i.e. state == 0 is the starting state?)